aten/src/ATen/native/AutogradComposite.cpp

```diff
@@ -117,6 +123,50 @@ static Tensor _lazy_clone_impl(Tensor const& self, bool future) {
   c10::intrusive_ptr<c10::StorageImpl> storage =
     c10::impl::cow::lazy_clone_storage(*self_storage, future);
 
+#if USE_MPS
+  // if (device == std::nullopt) {
+  //   std::cout << "Device is NON" << std::endl;
+  // } else if (device.value().is_mps()) {
+  //   std::cout << "Device is MPS" << std::endl;
+  // } else if (device.value().is_cpu()) {
+  //   std::cout << "Device is CPU" << std::endl;
+  // }
+  if (c10::impl::cow::get_future_lazy_clone()) {
+    // TODO(Frank): Unclear if this is the right approach.
+    if (device == std::nullopt) {
+      device = c10::Device("cpu");
+    }
+    at::Allocator* allocator = nullptr;
+    // if (device.value().is_cpu()) {
+    //   allocator = c10::GetDefaultCPUAllocator();
+    // } else if (device.value().is_mps()) {
+    //   allocator = at::mps::GetMPSAllocator();
+    // } else {
+    //   std::cout << "AutograComposite" << std::endl;
+    //   assert(false);
+    // }
+    // if (storage->allocator() == at::mps::GetMPSAllocator()) {
+    //   std::cout << "Changing from MPS Allocator to CPU Allocator" << std::endl;
+    // } else {
+    //   std::cout << "Changing from CPU Allocator to MPS Allocator" << std::endl;
+    // }
+
+    allocator = c10::GetAllocator(device.value().type());
+    if (allocator == nullptr) {
+      std::cout << "Allocator is nullptr" << std::endl;
+    } else if (allocator == at::mps::GetMPSAllocator()) {
+      std::cout << "Allocator is MPS allocator" << std::endl;
+    } else {
+      std::cout << "Allocator is CPU allocator" << std::endl;
+    }
+
+    storage->set_allocator(c10::GetAllocator(device.value().type()));
+    storage->_mutable_data_ptr_no_checks().unsafe_set_device(device.value());
```

Comment:
Maybe at first, try just setting the allocator and not the device. Does that at least fix the error you're getting?

Reply:
I would expect this to fail as well:

```python
import torch
torch.set_future_lazy_clone(True)

cpu = torch.zeros((1, 2, 3), device=torch.device('cpu'))
mps0 = cpu.to("mps")
mps1 = mps0._lazy_clone()
```